src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs

```diff
@@ -36,202 +79,565 @@ public SafeX509ChainHandle SafeHandle
+            Interop.Crypto.X509VerifyCert(storeCtx);
+            Interop.Crypto.X509VerifyStatusCode statusCode = Interop.Crypto.X509StoreCtxGetError(storeCtx);
 
-            // An X509_STORE is more comparable to Cryptography.X509Certificate2Collection than to
-            // Cryptography.X509Store. So read this with OpenSSL eyes, not CAPI/CNG eyes.
-            //
-            // (If you need to think of it as an X509Store, it's a volatile memory store)
-            using (SafeX509StoreHandle store = Interop.Crypto.X509StoreCreate())
-            using (SafeX509StoreCtxHandle storeCtx = Interop.Crypto.X509StoreCtxCreate())
-            using (SafeX509StackHandle extraCerts = Interop.Crypto.NewX509Stack())
+            if (IsCompleteChain(statusCode))
             {
-                Interop.Crypto.CheckValidOpenSslHandle(store);
-                Interop.Crypto.CheckValidOpenSslHandle(storeCtx);
-                Interop.Crypto.CheckValidOpenSslHandle(extraCerts);
+                return statusCode;
+            }
+
+            SafeX509StackHandle untrusted = _untrustedLookup;
+
+            if (extraCerts?.Count > 0)
+            {
+                foreach(X509Certificate2 cert in extraCerts)
+                {
+                    AddToStackAndUpRef(((OpenSslX509CertificateReader)cert.Pal).SafeHandle, untrusted);
+                }
+
+                Interop.Crypto.X509StoreCtxRebuildChain(storeCtx);
+                statusCode = Interop.Crypto.X509StoreCtxGetError(storeCtx);
+            }
+
+            return statusCode;
+        }
+
+        internal static bool IsCompleteChain(Interop.Crypto.X509VerifyStatusCode statusCode)
+        {
+            switch (statusCode)
+            {
+                case Interop.Crypto.X509VerifyStatusCode.X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT:
+                case Interop.Crypto.X509VerifyStatusCode.X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY:
+                    return false;
+                default:
+                    return true;
+            }
+        }
 
-                bool lookupCrl = revocationMode != X509RevocationMode.NoCheck;
+        internal Interop.Crypto.X509VerifyStatusCode FindChainViaAia(
+            ref List<X509Certificate2> downloadedCerts)
+        {
+            IntPtr lastCert = IntPtr.Zero;
+            SafeX509StoreCtxHandle storeCtx = _storeCtx;
+
+            Interop.Crypto.X509VerifyStatusCode statusCode =
+                Interop.Crypto.X509VerifyStatusCode.X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT;
 
-                foreach (X509Certificate2 cert in candidates)
+            while (!IsCompleteChain(statusCode))
+            {
+                using (SafeX509Handle currentCert = Interop.Crypto.X509StoreCtxGetCurrentCert(storeCtx))
                 {
-                    OpenSslX509CertificateReader pal = (OpenSslX509CertificateReader)cert.Pal;
+                    IntPtr currentHandle = currentCert.DangerousGetHandle();
```

Comment:
This feels a bit strange to me:
- X509StoreCtxGetCurrentCert is manufacturing a new SafeHandle, and will dispose of it
- But we're storing the IntPtr across that disposal boundary and comparing another handle against it

That suggests that disposing of the SafeHandle doesn't actually release the underlying IntPtr.  Which then begs the question why are we paying for the SafeHandle here?

Or have I misunderstood?

Reply:
The native call uprefs the native object, and the SafeHandle release calls free (downref).  The other value of the SafeHandle is type safety through the rest of the work here.